zookeeper-server/src/main/java/org/apache/zookeeper/server/NettyServerCnxn.java

```diff
@@ -411,19 +494,44 @@ public void receiveMessage(ChannelBuffer message) {
         }
     }
 
+    /**
+     * An event that triggers a change in the channel's "Auto Read" setting.
+     * Used for throttling. By using an enum we can treat the two values as
+     * singletons and compare with ==.
+     */
+    enum AutoReadEvent {
+        DISABLE,
+        ENABLE
+    }
+
     @Override
     public void disableRecv() {
-        disableRecvNoWait().awaitUninterruptibly();
+        if (throttled.compareAndSet(false, true)) {
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("Throttling - disabling recv {}", this);
+            }
+            channel.pipeline().fireUserEventTriggered(AutoReadEvent.DISABLE);
+        }
     }
-    
-    private ChannelFuture disableRecvNoWait() {
-        throttled = true;
-        if (LOG.isDebugEnabled()) {
-            LOG.debug("Throttling - disabling recv " + this);
+
+    /**
+     * Note that the netty implementation ignores the <code>waitDisableRecv</code>
```

Comment:
This comment is for the master branch, where there is a single `disableRecv(bool waitDisableRecv)` method. A more appropriate comment for 3.5 which has 2 methods might be something like:

"Note that both disableRecv() and disableRecvNoWait() are asynchronous in the netty implementation."

And put that comment on the `disableRecv()` method instead.

Reply:
Done.